storage/range_command.go

```diff
@@ -412,33 +438,29 @@ func (r *Range) InternalRangeLookup(batch engine.Engine, args *proto.InternalRan
 		// TODO(tschottdorf) Candidate for a ReplicaCorruptionError, once we
 		// introduce that.
 		if err = gogoproto.Unmarshal(kvs[i].Value.Bytes, &rds[i]); err != nil {
-			reply.SetGoError(err)
-			return nil
+			return reply, nil, err
 		}
 	}
-
 	reply.Ranges = rds
-	return intents
+
+	return reply, intents, nil
 }
 
 // InternalHeartbeatTxn updates the transaction status and heartbeat
 // timestamp after receiving transaction heartbeat messages from
 // coordinator. Returns the updated transaction.
-func (r *Range) InternalHeartbeatTxn(batch engine.Engine, ms *engine.MVCCStats,
-	args *proto.InternalHeartbeatTxnRequest, reply *proto.InternalHeartbeatTxnResponse) {
+func (r *Range) InternalHeartbeatTxn(batch engine.Engine, ms *engine.MVCCStats, args *proto.InternalHeartbeatTxnRequest) (reply proto.InternalHeartbeatTxnResponse, err error) {
 	key := keys.TransactionKey(args.Txn.Key, args.Txn.ID)
 
 	var txn proto.Transaction
-	ok, err := engine.MVCCGetProto(batch, key, proto.ZeroTimestamp, true, nil, &txn)
-	if err != nil {
-		reply.SetGoError(err)
-		return
-	}
-	// If no existing transaction record was found, initialize
-	// to the transaction in the request header.
-	if !ok {
-		gogoproto.Merge(&txn, args.Txn)
+	if ok, err := engine.MVCCGetProto(batch, key, proto.ZeroTimestamp, true, nil, &txn); err != nil {
+		return reply, err
+	} else if !ok {
+		// If no existing transaction record was found, initialize
+		// to the transaction in the request header.
+		txn = *args.Txn
```

Comment:
This is a shallow copy; the old version was a deep copy. I'd rather keep the `Merge` just to be safe.